### PR TITLE
mpfr: 4.2.0 -> 4.2.0-12

### PR DESCRIPTION
**Pull request by de11n**
## Description of changes

Adds support for updating the patches automatically via updateScript and runs the script on `mpfr` to apply 12 bugfix patches from upstream.

cc @trofi

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by trofi** [comment]
Looks nice! Thank you!

**Comment by de11n** [comment]
Who should I ping for approval?

**Comment by trofi** [comment]
https://discourse.nixos.org/ has a few `PRs ready for review` threads where you might be able to find `nixpkgs` committers to merge the changes. A few committers watch those threads closely.

**Comment by figsoda** [comment]
I don't know how to review this, so I posted it in the thread for you

File: pkgs/development/libraries/mpfr/default.nix [hunk 1]
```diff
@@ -9,23 +9,34 @@
...
-      "mirror://gnu/mpfr/${pname}-${version}.tar.xz"
+    urls = let name = "${finalAttrs.pname}-${getVersionNoPatches finalAttrs.version}"; in [
+      "https://www.mpfr.org/${name}/${name}.tar.xz"
+      "mirror://gnu/mpfr/${name}.tar.xz"
     ];
     hash = "sha256-BqN43xNQEkjBsttaqXeiyBJq6Emp2be+JUb7Spwm2ZM=";
   };
 
+  patches = builtins.genList
+    (i: builtins.toPath "${./patches}/patch${if i + 1 < 10 then "0" else ""}${toString (i + 1)}.patch")
```
Comment (amjoseph-nixpkgs) [inline pkgs/development/libraries/mpfr/default.nix:33]: The magic number "10" here is error-prone; people will forget to update it.

Could you please have the `./update.sh` script `echo $patch_count > num-patches.nix` and then `./import num-patches.nix`?

Or better yet, just write the actual `[ ./1.patch ./2.patch ./3.patch ]` into a `patches.nix` and import that.

**Review by ghost — COMMENTED** [review]
Er, why are we downloading all these patches instead of simply downloading the patched tarball?

If upstream doesn't post tarballs for these micro-releases then maybe either (a) they aren't important enough for us to track them (i.e. just wait for 4.2.1) or (b) we should use `fetchGit` instead of trying to fool around with patching tarballs.

**Pull request closed without merging**